karl/sandbox/ganache.py

```diff
@@ -0,0 +1,123 @@
+        "0x395df67f0c2d2d9fe1ad08d1bc8b6627011959b79c53d7dd6a3536a33ab8a4fd",
+        "0xe485d098507f54e7733a205420dfddbe58db035fa577fc294ebd14db90767a52",
+        "0xa453611d9419d0e56f499079478fd72c37b251a94bfde4d19872c44cf65386e3",
+        "0x829e924fdf021ba3dbbc4225edfece9aca04b929d6e75613329ca6f1d31c0bb4",
+        "0xb0057716d5917badaf911b193b12b910811c1497b5bada8d7711f758981c3773",
+    ]
+
+    def __init__(
+        self,
+        # Sandbox
+        internal_host="localhost",
+        internal_port=9545,
+        deterministic=True,
+        gas_price=1,
+        # Blockchain to fork
+        rpc=None,
+        block_number=None,
+        # Misc
+        verbosity=logging.NOTSET,
+    ):
+        # Blockchain to fork
+        self.rpc = rpc
+        self.block_number = block_number
+
+        # Where to start the sandbox
+        self.internal_host = internal_host
+        self.internal_port = internal_port
+        self.internal_rpc = "http://{host}:{port}".format(
+            host=self.internal_host, port=self.internal_port
+        )
+
+        # Gas price
+        self.gas_price = gas_price
+
+        # Account settings
+        self.deterministic = deterministic
+
+        # Verbosity
+        self.verbosity = verbosity
+
+        self.logger = logging.getLogger("Ganache")
+        self.logger.setLevel(self.verbosity)
+
+        # Building ganache-cli args
+        self.logger.debug("Setting up ganache-cli args")
+        args = ["ganache-cli"]
+
+        # Host
+        if self.internal_host is not None:
+            args.extend(["-h", self.internal_host])
+
+        # Port
+        if self.internal_port is not None:
+            args.extend(["-p", str(self.internal_port)])
+
+        # Deterministic
+        if self.deterministic:
+            args.append("-d")
+
+        # Gas price
+        args.extend(["-g", str(self.gas_price)])
+
+        # Fork blockchain
+        if self.rpc is not None:
+            if self.block_number is not None:
+                args.extend(
+                    [
+                        "-f",
+                        "{rpc}@{block_number}".format(
+                            rpc=self.rpc, block_number=block_number
+                        ),
+                    ]
+                )
+            else:
+                args.extend(["-f", "{rpc}".format(rpc=self.rpc)])
+
+        self.logger.debug("Starting ganache with\n{}".format(" ".join(args)))
+        self.process = subprocess.Popen(
+            args, shell=False, universal_newlines=True, stdout=subprocess.PIPE
+        )
+        self.logger.debug(self.process.args)
+        for l in self.process.stdout:
+            self.logger.debug(l)
+            # Wait for boot
+            if "Listening on" in str(l):
+                self.logger.debug(l)
+                break
+
+    def stop(self):
+        self.process.send_signal(signal.SIGTERM)
+
+
+class GanacheBaseException(Exception):
```

Comment:
We should put these into the exceptions file to keep things clean :slightly_smiling_face: We should also consider adding meaningful messages to the exceptions here, similar to the way proposed in the previous comments)

Reply:
Removed these exceptions completely.